### DISCUSSION TITLE
What outgoing requests does the api/auth/callback/<provider id> make?

### DISCUSSION THREAD

**Discussion by prandell** [nextauthjs/next-auth#4772] [Help]
### Question 💬

Hi guys, having a lot of trouble using the AzureAD provider in production.

The TL;DR / primary question is, **what is meant to happen when I access the callback url of my oauth provider?** 

Doing so directly locally brings me to my sign in page with an error saying the callback didn't contain any id_token. This sounds reasonable to me. However, in production, there seems to be some external request made, and the only information I'm given is:

```
CALLBACK_OAUTH_ERROR {
  "name": "OAuthCallbackError",
  "message": "outgoing request timed out after 40000ms",
  "stack": <trace>
}
```

So why could this be, with every configuration the same (using containers) aside from the host?

** LONG VERSION **

Locally, I am able to sign in using my AzureAD provider which directs me to microsoft to sign in with my corporate account, and I can see that the debug logs show:

```
CREATE_STATE {
  "state": "SUZoMRIzvJuyJxEsJZBj5UELQkMDEdzi-Baf6GQSj-Q",
  "maxAge": 900
}
GET_AUTHORIZATION_URL {
  "url": "https://login.microsoftonline.com/<tenant>/oauth2/v2.0/authorize?client_id=<client>&scope=openid%20profile%20email&response_type=code&redirect_uri=http%3A%2F%2Flocalhost%3A3000%2Fapi%2Fauth%2Fcallback%2Fazuread&state=SUZoMRIzvJuyJxEsJZBj5UELQkMDEdzi-Baf6GQSj-Q",
  "cookies": [
    {
      "name": "next-auth.state",
      "value": "<value>",
      "options": {
        "httpOnly": true,
        "sameSite": "lax",
        "path": "/",
        "secure": false,
        "expires": "2022-06-26T23:14:19.832Z"
      }
    }
  ]
}
```

You can see there is a step to create state. I have also tried using PKCE and there is a similar step to create the code challenge which is reflected in the authorization url

However, the app we have set up is obviously not at localhost, and the redirect_uri also expects to have the same production host. 

I am using docker containers, and the exact same image works locally at least to get us redirected to Microsoft and create the challenge, however in production I get these errors:

```
GET_AUTHORIZATION_URL_ERROR {
  "name": "RPError"
}
SIGNIN_OAUTH_ERROR {
  "error": {
    "name": "RPError"
  },
  "provider": {
    "id": "azuread",
    "name": "SSO",
    "type": "oauth",
    "wellKnown": "https://login.microsoftonline.com/<tenant>/v2.0/.well-known/openid-configuration",
    "authorization": {
      "params": {
        "scope": "openid profile email"
      }
    },
    "idToken": true,
    "checks": [
      "state"
    ],
    "tenantId": "<tenant>",
    "clientId": "<client>",
    "clientSecret": "<secret>",
    "httpOptions": {
      "timeout": 40000
    },
    "signinUrl": "https://<prod_host>/api/auth/signin/azuread",
    "callbackUrl": "https://<prod_host>/api/auth/callback/azuread"
  }
}
```

You can see that it fails to generate the code challenge and auth url despite being the exact same code. Additionally, the reason I increased the http timeout is because it actually takes the entire duration of this timeout before showing this error, which I similarly don't quite understand

Any help would be appreciated, this has been driving me mad!

### How to reproduce ☕️

Unfortunately this cannot be reproduced as it may be specific to my corporate set up, however the steps are roughly

However my api/auth/[...nextauth].js file is as such:

```
export default NextAuth({
	debug: true,
	logger: {
		error(code, metadata) {
			console.log(code, JSON.stringify(metadata, null, 2))
		},
		warn(code) {
			console.log(code)
		},
		debug(code, metadata) {
			console.log(code, JSON.stringify(metadata, null, 2))
		}
	},
	
	providers: [
		AzureADProvider({
			id: 'azuread',
			name: 'SSO',
			tenantId: process.env.AZURE_AD_B2C_TENANT_ID,
			clientId: process.env.AZURE_AD_B2C_CLIENT_ID,
			clientSecret: process.env.AZURE_AD_B2C_CLIENT_SECRET,
			checks: ['pkce'],
			httpOptions: {
				timeout: 40000
			}
		}),
	],
	session: {
		strategy: 'jwt',
		maxAge: 24 * 60 * 60
	},
	pages: {
		signIn: '/login',
		error: '/login'
	}
})
```

### Contributing 🙌🏽

Yes, I am willing to help answer this question in a PR

**Comment by balazsorban44** [comment]
The general OAuth flow is shown here https://next-auth.js.org/configuration/providers/oauth

There might be small differences based on how you set up the sign-in process and how you configured your provider though, but should give you the idea.

> Unfortunately this cannot be reproduced as it may be specific to my corporate set up

This makes it very hard to recommend anything specific. 😕

To me it sounds like a proxy issue, maybe requests to your provider are blocked.

**Reply by prandell** [reply]
Thanks for the prompt response!

I have stripped it down to the most basic example of AzureAD provider, simply providing the 3 requested values from the config. I have also removed all my callbacks and page overrides and attempted to login by the generated sign in page.

Again, it simply times out when I attempt to log in without ever redirecting to the provider.

My initial suspicion was also proxy related issues, however when I `exec` into the container and curl the providers well-known configuration endpoint, I indeed receive a response. So it's not a docker proxy issue.

However if it is a proxy issue at all, I'm not even sure how to determine which particular call from next-auth is failing? Is there a way I can determine that?

**Reply by prandell** [reply]
I did come across this https://github.com/nextauthjs/next-auth/pull/3931/files?short_path=32b817d#diff-32b817d3359da07a21167689d05fe2caa740ea9eb2d2005697ee35c7d5be948a

which for a while had me convinced was the issue. However, when running the exact same container locally I have no problems being redirected to the provider after generating the authorization url. All requests can be made manually from outside the container and within it in production as well

**Comment by prandell** [comment]
If it would be possible to get more logs or further insight as to what is going wrong at the steps highlighted, it may provide me with enough insight to determine what the problem is

All I can tell is that there is a request timing out, but I can't see what **exactly** that request is or if we even managed to compute the `authorisation_url` with the correct params

![Screen Shot 2022-06-27 at 11 42 45 am](https://user-images.githubusercontent.com/52766847/175845095-bbe020f1-4aee-4632-a939-39777406733b.png)

**Comment by prandell** [comment]
I followed the tutorial here and made the changes to allow for a proxy variable https://github.com/nextauthjs/next-auth/pull/3931/files?short_path=32b817d#diff-32b817d3359da07a21167689d05fe2caa740ea9eb2d2005697ee35c7d5be948a

However, now my error messages are simply:

```
GET_AUTHORIZATION_URL_ERROR {
  "code": "ERR_INVALID_PROTOCOL"
}
SIGNIN_OAUTH_ERROR {
  "error": {
    "code": "ERR_INVALID_PROTOCOL"
  },
  "provider": {
    "id": "azure-ad",
    "name": "Azure Active Directory",
    "type": "oauth",
    "wellKnown": "https://login.microsoftonline.com/<tenant>/v2.0/.well-known/openid-configuration",
    "authorization": {
      "params": {
        "scope": "openid profile email"
      }
    },
    "idToken": true,
    "checks": [
      "pkce"
    ],
    "tenantId": "<tenant>",
    "clientId": "<client>",
    "clientSecret": "<secret>",
    "signinUrl": "<host>/api/auth/signin/azure-ad",
    "callbackUrl": "<host>/api/auth/callback/azure-ad"
  }
}
```

There doesn't seem to be any documentation on this error either

**Reply by ginibrandonnpx** [reply]
Did you ever find a solution to this? I'm having the same symptoms.
